Rust_Scylla_Driver/notes/restserver/src/main.rs

```diff
@@ -0,0 +1,243 @@
+        payload.topic.to_string(),
+        payload.content.to_string(),
+        permit,
+    )));
+
+    debug!("Waiting for notes to be processed...");
+    for thread in handlers {
+        match thread.await {
+            Err(e) => return Err(ErrorInternalServerError(e)),
+            Ok(r) => {
+                if let Err(e) = r {
+                    error!("Error: {:?}", e);
+                    return Err(ErrorInternalServerError(e));
+                }
+            }
+        }
+    }
+
+    let elapsed = now.elapsed();
+    info!("Ingestion Time: {:.2?}", elapsed);
+    Ok(HttpResponse::Ok().json(r#"{ "status": "OK"}"#))
+}
+
+async fn add_note(
+    id: Uuid,
+    state: Data<AppState>,
+    topic: String,
+    content: String,
+    permit: Result<OwnedSemaphorePermit, AcquireError>,
+) -> Result<(), Box<dyn std::error::Error + Sync + Send>> {
+    let ltopic= topic.clone();
+    info!(
+        "Processing note {} with ID {}. ",
+        ltopic, id
+    );
+    let now = Instant::now();
+    // persist in DB and wait
+    state.db_svc.save_note(DbNote{id,topic,content}).await?;
+
+    let elapsed = now.elapsed();
+    info!("Note {} processed. Took {:.2?}", ltopic, elapsed);
+
+    let _permit = permit;
+
+    Ok(())
+}
+
+#[actix_web::main]
+async fn main() -> Result<()> {
+    let config = Config::from_env().expect("Server configuration");
+
+    let port = config.port;
+    let host = config.host.clone();
+    let num_cpus = num_cpus::get();
+    let parallel_inserts = config.parallel_inserts;
+    let db_parallelism = config.db_parallelism;
+
+    info!(
+        "Starting application. Num CPUs {}. Max Parallel Files {}. DB Parallelism {}. ",
+        num_cpus, parallel_inserts, db_parallelism
+    );
+
+    let db = ScyllaDbService::new(config.db_dc, config.db_url,
+        config.db_user, config.db_password,
+        db_parallelism, config.schema_file).await;
+
+    let sem = Arc::new(Semaphore::new(parallel_inserts));
+    let data = Data::new(AppState {
+        db_svc: db,
+        semaphore: sem
+    });
+
+    info!("Starting server at http://{}:{}/", host, port);
+//     let cors = Cors::default() //if you want production safe code, you need to send Cors headers in your server
+// //        .allowed_origin("http://localhost")
+//         .send_wildcard()
+//         .allowed_methods(vec!["GET", "POST", "DELETE"])
+//         .allowed_headers(vec![header::AUTHORIZATION, header::ACCEPT])
+//         .allowed_header(header::CONTENT_TYPE)
+//         .max_age(3600);
+
+    HttpServer::new(move || {
+        let cors_develop = Cors::permissive(); //temporary development trick
+        App::new()
+            .wrap(cors_develop)
+            .wrap(Logger::default())
+            .app_data(data.clone())
+            .service(insert_note)
+            .service(get_all_notes)
+            .service(get_by_id)
+            .service(delete_by_id)
+    })
+    .bind(format!("{}:{}", host, port))?
+    .workers(num_cpus * 2)
```

Comment:
```
let num_cpus = num_cpus::get();
```
According to the documentation, the `num_cpus` is the number of logical cores and not physical, so I guess multiplying by 2 is redundant here.

Reply:
the idea was that this is performance tuning, since some workers will be waiting on IO to Scylla when processing requests
so the question is how well does the async work here with actix workers ...